File: apps/go-sdk/agent_test.go
package firecrawl

import (
	"context"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"

	"github.com/firecrawl/firecrawl/apps/go-sdk/option"
)

// capturedRequest ferries the request the httptest handler received to the
// test goroutine over a channel. Plain shared variables would trip the race
// detector (the handler runs on its own goroutine), and t.Fatal inside the
// handler would Goexit without writing a response, surfacing as a connection
// error instead of the assertion message.
type capturedRequest struct {
	method string
	path   string
	query  string
	body   string
}

func captureRequest(r *http.Request) capturedRequest {
	c := capturedRequest{method: r.Method, path: r.URL.Path, query: r.URL.RawQuery}
	if r.Body != nil {
		data, _ := io.ReadAll(r.Body)
		c.body = string(data)
	}
	return c
}

func TestStartAgentSendsEffort(t *testing.T) {
	captured := make(chan capturedRequest, 1)

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		captured <- captureRequest(r)

		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte(`{"success":true,"id":"job-123"}`))
	}))
	defer server.Close()

	client, err := NewClient(
		option.WithAPIKey("fc-test"),
		option.WithAPIURL(server.URL),
	)
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}

	resp, err := client.StartAgent(context.Background(), &AgentOptions{
		Prompt: "find pricing",
		Model:  String("spark-2"),
		Effort: String("high"),
	})
	if err != nil {
		t.Fatalf("StartAgent: %v", err)
	}

	req := <-captured
	if req.method != http.MethodPost || req.path != "/v2/agent" {
		t.Errorf("unexpected request: %s %s", req.method, req.path)
	}
	if resp.ID != "job-123" {
		t.Errorf("id = %q, want job-123", resp.ID)
	}
	if !strings.Contains(req.body, `"effort":"high"`) {
		t.Errorf("request body missing effort: %q", req.body)
	}
	if !strings.Contains(req.body, `"model":"spark-2"`) {
		t.Errorf("request body missing model: %q", req.body)
	}
}

func TestStartAgentOmitsEffortWhenUnset(t *testing.T) {
	captured := make(chan capturedRequest, 1)

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		captured <- captureRequest(r)

		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte(`{"success":true,"id":"job-123"}`))
	}))
	defer server.Close()

	client, err := NewClient(
		option.WithAPIKey("fc-test"),
		option.WithAPIURL(server.URL),
	)
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}

	if _, err := client.StartAgent(context.Background(), &AgentOptions{Prompt: "find pricing"}); err != nil {
		t.Fatalf("StartAgent: %v", err)
	}

	req := <-captured
	if strings.Contains(req.body, `"effort"`) {
		t.Errorf("request body should omit effort: %q", req.body)
	}
}

func TestGetAgentTraceParsesEvents(t *testing.T) {
	captured := make(chan capturedRequest, 1)

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		captured <- captureRequest(r)

		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte(`{
			"success": true,
			"id": "job-123",
			"creditsUsed": 5,
			"events": [
				{
					"type": "run.started",
					"schemaVersion": 1,
					"eventId": "evt-1",
					"runId": "job-123",
					"occurredAt": "2026-08-26T10:00:00Z",
					"producerSequence": 1,
					"agent": {"id": "agent-1", "role": "primary", "name": "spark-2"}
				},
				{
					"type": "artifact.updated",
					"schemaVersion": 1,
					"eventId": "evt-2",
					"runId": "job-123",
					"occurredAt": "2026-08-26T10:00:05Z",
					"producerSequence": 2,
					"agent": {"id": "agent-1", "role": "primary", "name": "spark-2"},
					"artifact": {
						"kind": "json",
						"artifactId": "art-1",
						"path": "plans",
						"snapshotId": "snap-1",
						"change": "updated",
						"changedFields": ["plans"],
						"itemCount": 3,
						"sourceToolCallId": "tc-1"
					}
				}
			]
		}`))
	}))
	defer server.Close()

	client, err := NewClient(
		option.WithAPIKey("fc-test"),
		option.WithAPIURL(server.URL),
	)
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}

	trace, err := client.GetAgentTrace(context.Background(), "job-123", false)
	if err != nil {
		t.Fatalf("GetAgentTrace: %v", err)
	}

	req := <-captured
	if req.method != http.MethodGet || req.path != "/v2/agent/job-123/trace" {
		t.Errorf("unexpected request: %s %s", req.method, req.path)
	}
	if req.query != "" {
		t.Errorf("unexpected query params: %q", req.query)
	}
	if !trace.Success || trace.ID != "job-123" {
		t.Fatalf("trace = %+v", trace)
	}
	if trace.CreditsUsed == nil || *trace.CreditsUsed != 5 {
		t.Fatalf("creditsUsed = %v, want 5", trace.CreditsUsed)
	}
	if len(trace.Events) != 2 {
		t.Fatalf("events len = %d, want 2", len(trace.Events))
	}

	started := trace.Events[0]
	if started.Type != "run.started" || started.EventID != "evt-1" || started.RunID != "job-123" {
		t.Fatalf("run.started event = %+v", started)
	}
	if started.SchemaVersion != 1 || started.ProducerSequence != 1 || started.OccurredAt != "2026-08-26T10:00:00Z" {
		t.Fatalf("run.started base fields = %+v", started)
	}
	if started.Agent.ID != "agent-1" || started.Agent.Role != "primary" || started.Agent.Name != "spark-2" {
		t.Fatalf("run.started agent = %+v", started.Agent)
	}

	updated := trace.Events[1]
	if updated.Type != "artifact.updated" || updated.Artifact == nil {
		t.Fatalf("artifact.updated event = %+v", updated)
	}
	artifact := updated.Artifact
	if artifact.Kind != "json" || artifact.ArtifactID != "art-1" || artifact.SnapshotID != "snap-1" || artifact.Change != "updated" {
		t.Fatalf("artifact = %+v", artifact)
	}
	if artifact.Path != "plans" || artifact.SourceToolCallID != "tc-1" {
		t.Fatalf("artifact = %+v", artifact)
	}
	if len(artifact.ChangedFields) != 1 || artifact.ChangedFields[0] != "plans" {
		t.Fatalf("changedFields = %v", artifact.ChangedFields)
	}
	if artifact.ItemCount == nil || *artifact.ItemCount != 3 {
		t.Fatalf("itemCount = %v, want 3", artifact.ItemCount)
	}
}

func TestGetAgentTraceSendsLiveView(t *testing.T) {
	captured := make(chan capturedRequest, 1)

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		captured <- captureRequest(r)

		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte(`{
			"success": true,
			"id": "job-123",
			"events": [],
			"activeBrowserSessions": [
				{
					"id": "session-1",
					"liveViewUrl": "https://live.firecrawl.dev/session-1",
					"viewport": {"width": 1280, "height": 720}
				}
			]
		}`))
	}))
	defer server.Close()

	client, err := NewClient(
		option.WithAPIKey("fc-test"),
		option.WithAPIURL(server.URL),
	)
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}

	trace, err := client.GetAgentTrace(context.Background(), "job-123", true)
	if err != nil {
		t.Fatalf("GetAgentTrace: %v", err)
	}

	req := <-captured
	if req.method != http.MethodGet || req.path != "/v2/agent/job-123/trace" {
		t.Errorf("unexpected request: %s %s", req.method, req.path)
	}
	if req.query != "liveView=true" {
		t.Errorf("liveView query param = %q, want true", req.query)
	}
	if len(trace.ActiveBrowserSessions) != 1 {
		t.Fatalf("activeBrowserSessions len = %d, want 1", len(trace.ActiveBrowserSessions))
	}
	session := trace.ActiveBrowserSessions[0]
	if session.ID != "session-1" || session.LiveViewURL != "https://live.firecrawl.dev/session-1" {
		t.Fatalf("session = %+v", session)
	}
	if session.Viewport.Width != 1280 || session.Viewport.Height != 720 {
		t.Fatalf("viewport = %+v", session.Viewport)
	}
}

func TestGetAgentTraceRequiresJobID(t *testing.T) {
	client, err := NewClient(
		option.WithAPIKey("fc-test"),
		option.WithAPIURL("http://localhost:0"),
	)
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}

	if _, err := client.GetAgentTrace(context.Background(), "", false); err == nil {
		t.Fatalf("expected error for empty job ID")
	}
}

func TestGetAgentSnapshotParsesResponse(t *testing.T) {
	captured := make(chan capturedRequest, 1)

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		captured <- captureRequest(r)

		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte(`{
			"success": true,
			"id": "job-123",
			"snapshotId": "snap-1",
			"snapshot": "{\"plans\":[{\"name\":\"pro\",\"price\":\"$20\"}]}"
		}`))
	}))
	defer server.Close()

	client, err := NewClient(
		option.WithAPIKey("fc-test"),
		option.WithAPIURL(server.URL),
	)
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}

	snapshot, err := client.GetAgentSnapshot(context.Background(), "job-123", "snap-1")
	if err != nil {
		t.Fatalf("GetAgentSnapshot: %v", err)
	}

	req := <-captured
	if req.method != http.MethodGet || req.path != "/v2/agent/job-123/snapshots/snap-1" {
		t.Errorf("unexpected request: %s %s", req.method, req.path)
	}
	if !snapshot.Success || snapshot.ID != "job-123" || snapshot.SnapshotID != "snap-1" {
		t.Fatalf("snapshot = %+v", snapshot)
	}
	if !strings.Contains(snapshot.Snapshot, `"plans"`) {
		t.Errorf("snapshot content = %q", snapshot.Snapshot)
	}
}

func TestGetAgentSnapshotRequiresIDs(t *testing.T) {
	client, err := NewClient(
		option.WithAPIKey("fc-test"),
		option.WithAPIURL("http://localhost:0"),
	)
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}

	if _, err := client.GetAgentSnapshot(context.Background(), "", "snap-1"); err == nil {
		t.Fatalf("expected error for empty job ID")
	}
	if _, err := client.GetAgentSnapshot(context.Background(), "job-123", ""); err == nil {
		t.Fatalf("expected error for empty snapshot ID")
	}
}

File: apps/php-sdk/tests/Unit/AgentTest.php
<?php

declare(strict_types=1);

use Firecrawl\Exceptions\FirecrawlException;
use Firecrawl\Models\AgentOptions;
use Firecrawl\Models\AgentStatusResponse;
use GuzzleHttp\Psr7\Response;

it('sends the effort parameter when starting an agent', function (): void {
    $history = new ArrayObject();
    $client = fakeFirecrawlClient([
        new Response(200, [], json_encode(['success' => true, 'id' => 'agent-123'])),
    ], $history);

    $response = $client->startAgent(AgentOptions::with(
        prompt: 'find pricing',
        effort: 'high',
    ));

    expect($response->isSuccess())->toBeTrue();
    expect($response->getId())->toBe('agent-123');

    $request = $history[0]['request'];
    expect($request->getUri()->getPath())->toBe('/v2/agent');

    $body = json_decode((string) $request->getBody(), true);
    expect($body['prompt'])->toBe('find pricing');
    expect($body['effort'])->toBe('high');
});

it('omits the effort parameter when not set', function (): void {
    $history = new ArrayObject();
    $client = fakeFirecrawlClient([
        new Response(200, [], json_encode(['success' => true, 'id' => 'agent-123'])),
    ], $history);

    $client->startAgent(AgentOptions::with(prompt: 'find pricing', model: 'spark-2'));

    $body = json_decode((string) $history[0]['request']->getBody(), true);
    expect($body)->not->toHaveKey('effort');
    expect($body['model'])->toBe('spark-2');
});

it('hydrates the effort on AgentStatusResponse', function (): void {
    $withEffort = AgentStatusResponse::fromArray([
        'success' => true,
        'status' => 'completed',
        'model' => 'spark-2',
        'effort' => 'medium',
        'expiresAt' => '2026-09-01T00:00:00Z',
    ]);

    expect($withEffort->getEffort())->toBe('medium');
    expect($withEffort->getModel())->toBe('spark-2');

    $withoutEffort = AgentStatusResponse::fromArray([
        'success' => true,
        'status' => 'completed',
    ]);

    expect($withoutEffort->getEffort())->toBeNull();
});

it('parses agent trace events and sends the liveView query param', function (): void {
    $history = new ArrayObject();
    $client = fakeFirecrawlClient([
        new Response(200, [], json_encode([
            'success' => true,
            'id' => 'agent-123',
            'events' => [
                [
                    'type' => 'run.started',
                    'schemaVersion' => 1,
                    'eventId' => 'evt-1',
                    'runId' => 'agent-123',
                    'occurredAt' => '2026-08-26T10:00:00Z',
                    'producerSequence' => 1,
                    'agent' => ['id' => 'agent-root', 'role' => 'root', 'name' => 'root', 'parentId' => null],
                ],
                [
                    'type' => 'tool_call.started',
                    'schemaVersion' => 1,
                    'eventId' => 'evt-2',
                    'runId' => 'agent-123',
                    'occurredAt' => '2026-08-26T10:00:01Z',
                    'producerSequence' => 2,
                    'agent' => ['id' => 'agent-root', 'role' => 'root', 'name' => 'root'],
                    'toolCallId' => 'call-1',
                    'toolName' => 'scrape',
                    'parameters' => ['url' => 'https://example.com'],
                ],
                [
                    'type' => 'artifact.updated',
                    'schemaVersion' => 1,
                    'eventId' => 'evt-3',
                    'runId' => 'agent-123',
                    'occurredAt' => '2026-08-26T10:00:02Z',
                    'producerSequence' => 3,
                    'agent' => ['id' => 'agent-root', 'role' => 'root', 'name' => 'root'],
                    'artifact' => [
                        'kind' => 'table',
                        'artifactId' => 'art-1',
                        'path' => 'pricing',
                        'snapshotId' => 'snap-1',
                        'change' => 'rows_added',
                        'changedFields' => ['price'],
                        'itemCount' => 3,
                        'sourceToolCallId' => 'call-1',
                    ],
                ],
                [
                    'type' => 'error.occurred',
                    'schemaVersion' => 1,
                    'eventId' => 'evt-4',
                    'runId' => 'agent-123',
                    'occurredAt' => '2026-08-26T10:00:03Z',
                    'producerSequence' => 4,
                    'agent' => ['id' => 'agent-root', 'role' => 'root', 'name' => 'root'],
                    'error' => [
                        'code' => 'SCRAPE_FAILED',
                        'source' => 'tool',
                        'retryable' => true,
                        'message' => 'scrape timed out',
                    ],
                ],
            ],
            'creditsUsed' => 5,
            'activeBrowserSessions' => [
                [
                    'id' => 'session-1',
                    'liveViewUrl' => 'https://liveview.firecrawl.dev/session-1',
                    'viewport' => ['width' => 1280, 'height' => 720],
                ],
            ],
        ])),
    ], $history);

    $trace = $client->getAgentTrace('agent-123', liveView: true);

    $request = $history[0]['request'];
    expect($request->getMethod())->toBe('GET');
    expect($request->getUri()->getPath())->toBe('/v2/agent/agent-123/trace');
    expect($request->getUri()->getQuery())->toBe('liveView=true');

    expect($trace->isSuccess())->toBeTrue();
    expect($trace->getId())->toBe('agent-123');
    expect($trace->getCreditsUsed())->toBe(5);
    expect($trace->getEvents())->toHaveCount(4);

    $started = $trace->getEvents()[0];
    expect($started->getType())->toBe('run.started');
    expect($started->getSchemaVersion())->toBe(1);
    expect($started->getEventId())->toBe('evt-1');
    expect($started->getRunId())->toBe('agent-123');
    expect($started->getOccurredAt())->toBe('2026-08-26T10:00:00Z');
    expect($started->getProducerSequence())->toBe(1);
    expect($started->getAgent()->getId())->toBe('agent-root');
    expect($started->getAgent()->getRole())->toBe('root');
    expect($started->getAgent()->getParentId())->toBeNull();
    expect($started->getError())->toBeNull();
    expect($started->getArtifact())->toBeNull();

    $toolCall = $trace->getEvents()[1];
    expect($toolCall->getType())->toBe('tool_call.started');
    expect($toolCall->getToolCallId())->toBe('call-1');
    expect($toolCall->getToolName())->toBe('scrape');
    expect($toolCall->getParameters())->toBe(['url' => 'https://example.com']);
    expect($toolCall->getResult())->toBeNull();

    $artifact = $trace->getEvents()[2]->getArtifact();
    expect($artifact->getKind())->toBe('table');
    expect($artifact->getArtifactId())->toBe('art-1');
    expect($artifact->getPath())->toBe('pricing');
    expect($artifact->getSnapshotId())->toBe('snap-1');
    expect($artifact->getChange())->toBe('rows_added');
    expect($artifact->getChangedFields())->toBe(['price']);
    expect($artifact->getItemCount())->toBe(3);
    expect($artifact->getSourceToolCallId())->toBe('call-1');

    $error = $trace->getEvents()[3]->getError();
    expect($error->getCode())->toBe('SCRAPE_FAILED');
    expect($error->getSource())->toBe('tool');
    expect($error->isRetryable())->toBeTrue();
    expect($error->getMessage())->toBe('scrape timed out');

    $sessions = $trace->getActiveBrowserSessions();
    expect($sessions)->toHaveCount(1);
    expect($sessions[0]->getId())->toBe('session-1');
    expect($sessions[0]->getLiveViewUrl())->toBe('https://liveview.firecrawl.dev/session-1');
    expect($sessions[0]->getViewport()->getWidth())->toBe(1280);
    expect($sessions[0]->getViewport()->getHeight())->toBe(720);
});

it('omits the liveView query param by default', function (): void {
    $history = new ArrayObject();
    $client = fakeFirecrawlClient([
        new Response(200, [], json_encode([
            'success' => true,
            'id' => 'agent-123',
            'events' => [],
            'creditsUsed' => 0,
        ])),
    ], $history);

    $trace = $client->getAgentTrace('agent-123');

    expect($history[0]['request']->getUri()->getQuery())->toBe('');
    expect($trace->getEvents())->toBe([]);
    expect($trace->getActiveBrowserSessions())->toBe([]);
});

it('surfaces agent trace errors', function (): void {
    $client = fakeFirecrawlClient([
        new Response(404, [], json_encode([
            'success' => false,
            'error' => 'Agent job not found',
        ])),
    ]);

    $client->getAgentTrace('agent-missing');
})->throws(FirecrawlException::class, 'Agent job not found');

it('parses an agent snapshot', function (): void {
    $history = new ArrayObject();
    $client = fakeFirecrawlClient([
        new Response(200, [], json_encode([
            'success' => true,
            'id' => 'agent-123',
            'snapshotId' => 'snap-1',
            'snapshot' => '[{"price": "$10"}]',
        ])),
    ], $history);

    $snapshot = $client->getAgentSnapshot('agent-123', 'snap-1');

    $request = $history[0]['request'];
    expect($request->getMethod())->toBe('GET');
    expect($request->getUri()->getPath())->toBe('/v2/agent/agent-123/snapshots/snap-1');

    expect($snapshot->isSuccess())->toBeTrue();
    expect($snapshot->getId())->toBe('agent-123');
    expect($snapshot->getSnapshotId())->toBe('snap-1');
    expect($snapshot->getSnapshot())->toBe('[{"price": "$10"}]');
    expect($snapshot->getError())->toBeNull();
});

it('surfaces agent snapshot errors', function (): void {
    $client = fakeFirecrawlClient([
        new Response(404, [], json_encode([
            'success' => false,
            'error' => 'Agent job not found',
        ])),
    ]);

    $client->getAgentSnapshot('agent-missing', 'snap-1');
})->throws(FirecrawlException::class, 'Agent job not found');
